nifi-extension-bundles/nifi-aws-bundle/nifi-aws-kinesis/src/main/java/org/apache/nifi/processors/aws/kinesis/ConsumeKinesis.java

```diff
@@ -0,0 +1,737 @@
+    private static InitialPositionInStreamExtended getInitialPosition(final ProcessContext context) {
+        final InitialPosition initialPosition = context.getProperty(INITIAL_STREAM_POSITION).asAllowableValue(InitialPosition.class);
+        return switch (initialPosition) {
+            case TRIM_HORIZON ->
+                    InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON);
+            case LATEST -> InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.LATEST);
+            case AT_TIMESTAMP -> {
+                final String timestampValue = context.getProperty(STREAM_POSITION_TIMESTAMP).getValue();
+                final Instant timestamp = Instant.parse(timestampValue);
+                yield InitialPositionInStreamExtended.newInitialPositionAtTimestamp(Date.from(timestamp));
+            }
+        };
+    }
+
+    private static @Nullable MetricsFactory configureMetricsFactory(final ProcessContext context) {
+        final MetricsPublishing metricsPublishing = context.getProperty(METRICS_PUBLISHING).asAllowableValue(MetricsPublishing.class);
+        return switch (metricsPublishing) {
+            case DISABLED -> new NullMetricsFactory();
+            case LOGS -> new LogMetricsFactory();
+            case CLOUDWATCH -> null; // If no metrics factory was provided, CloudWatch metrics factory is used by default.
+        };
+    }
+
+    @OnStopped
+    public void onStopped() {
+        if (kinesisScheduler != null) {
+            shutdownScheduler();
+            kinesisScheduler = null;
+        }
+        if (kinesisClient != null) {
+            kinesisClient.close();
+            kinesisClient = null;
+        }
+        if (dynamoDbClient != null) {
+            dynamoDbClient.close();
+            dynamoDbClient = null;
+        }
+        if (cloudWatchClient != null) {
+            cloudWatchClient.close();
+            cloudWatchClient = null;
+        }
+
+        recordBuffer = null;
+        readerRecordProcessor = Optional.empty();
+    }
+
+    private void shutdownScheduler() {
+        final long start = System.nanoTime();
+        getLogger().debug("Shutting down Kinesis Scheduler");
+
+        boolean gracefulShutdownSucceeded;
+        try {
+            gracefulShutdownSucceeded = kinesisScheduler.startGracefulShutdown().get(KINESIS_SCHEDULER_GRACEFUL_SHUTDOWN_TIMEOUT.getSeconds(), SECONDS);
+            if (!gracefulShutdownSucceeded) {
+                getLogger().warn("Failed to shutdown Kinesis Scheduler gracefully. See the logs for more details");
+            }
+        } catch (final RuntimeException | InterruptedException | ExecutionException | TimeoutException e) {
+            getLogger().warn("Failed to shutdown Kinesis Scheduler gracefully", e);
+            gracefulShutdownSucceeded = false;
+        }
+
+        if (!gracefulShutdownSucceeded) {
+            getLogger().warn("Falling back to a forceful shutdown of Kinesis Scheduler");
```

Comment:
This log seems unnecessary given the other warnings about graceful shutdown failing.